# 1 Objetivo
Crear y evaluar un modelo de arboles aleatotios (random forest) para predecir las ventas con datos simulados de una empresa dependiendo de las inversiones realizadas en publicidad.

# 2 Descripción
- Se utiliza programación Python
- Cargar librerías y datos
- Limpiar datos si es necesario
- Explorar datos
- Partir los datos en datos de entrenamiento y datos de validación 70% y 30%
- Crear modelo de regresión con los datos de entrenamiento
- Predicciones
- Evaluar predicciones con respecto a datos reales
- Determinar el estadístico rmse para evaluar con respecto a otros modelos
- Interpretar el caso

# 3 Fundamento teórico
Extraído de : (Amat Rodrigo 2017)

Un modelo Random Forest está formado por un conjunto (ensemble) de árboles de decisión individuales, cada uno entrenado con una muestra aleatoria extraída de los datos de entrenamiento originales mediante bootstrapping. Esto implica que cada árbol se entrena con unos datos ligeramente distintos.

En cada árbol individual, las observaciones se van distribuyendo por bifurcaciones (nodos) generando la estructura del árbol hasta alcanzar un nodo terminal. La predicción de una nueva observación se obtiene agregando las predicciones de todos los árboles individuales que forman el modelo.

Para entender cómo funcionan los modelos Random Forest es necesario conocer primero los conceptos de ensemble y bagging.

Métodos de ensemble

Todos los modelos de aprendizaje estadístico y machine learning sufren el problema de equilibrio entre bias y varianza.

El término bias (sesgo) hace referencia a cuánto se alejan en promedio las predicciones de un modelo respecto a los valores reales. Refleja cómo de capaz es el modelo de aprender la relación real que existe entre los predictores y la variable respuesta. Por ejemplo, si la relación sigue un patrón no lineal, por muchos datos de los que se disponga, un modelo de regresión lineal no podrá modelar correctamente la relación, por lo que tendrá un bias alto.

El término varianza hace referencia a cuánto cambia el modelo dependiendo de los datos utilizados en su entrenamiento. Idealmente, un modelo no debería modificarse demasiado por pequeñas variaciones en los datos de entrenamiento, si esto ocurre, es porque el modelo está memorizando los datos en lugar de aprender la verdadera relación entre los predictores y la variable respuesta. Por ejemplo, un modelo de árbol con muchos nodos, suele variar su estructura con que apenas cambien unos pocos datos de entrenamiento, tiene mucha varianza.

A medida que aumenta la complejidad de un modelo, este dispone de mayor flexibilidad para adaptarse a las observaciones, reduciendo así el bias y mejorando su capacidad predictiva. Sin embargo, alcanzado un determinado grado de flexibilidad, aparece el problema de overfitting, el modelo se ajusta tanto a los datos de entrenamiento que es incapaz de predecir correctamente nuevas observaciones. El mejor modelo es aquel que consigue un equilibrio óptimo entre bias y varianza.

¿Cómo se controlan el bias y varianza en los modelos basados en árboles? Por lo general, los árboles pequeños (pocas ramificaciones) tienen poca varianza pero no consiguen representar bien la relación entre las variables, es decir, tienen bias alto. En contraposición, los árboles grandes se ajustan mucho a los datos de entrenamiento, por lo que tienen muy poco bias pero mucha varianza. Una forma de solucionar este problema son los métodos de ensemble.

Los métodos de ensemble combinan múltiples modelos en uno nuevo con el objetivo de lograr un equilibro entre bias y varianza, consiguiendo así mejores predicciones que cualquiera de los modelos individuales originales. Dos de los tipos de ensemble más utilizados son:

Bagging: Se ajustan múltiples modelos, cada uno con un subconjunto distinto de los datos de entrenamiento. Para predecir, todos los modelos que forman el agregado participan aportando su predicción. Como valor final, se toma la media de todas las predicciones (variables continuas) o la clase más frecuente (variables categóricas). Los modelos Random Forest están dentro de esta categoría.

Boosting: Se ajustan secuencialmente múltiples modelos sencillos, llamados weak learners, de forma que cada modelo aprende de los errores del anterior. Como valor final, al igual que en bagging, se toma la media de todas las predicciones (variables continuas) o la clase más frecuente (variables cualitativas). Tres de los métodos de boosting más empleados son AdaBoost, Gradient Boosting y Stochastic Gradient Boosting.

Aunque el objetivo final es el mismo, lograr un balance óptimo entre bias y varianza, existen dos diferencias importantes:

Forma en que consiguen reducir el error total. El error total de un modelo puede descomponerse como bias+varianza+ϵ.

En bagging, se emplean modelos con muy poco bias pero mucha varianza, agregándolos se consigue reducir la varianza sin apenas inflar el bias. En boosting, se emplean modelos con muy poca varianza pero mucho bias, ajustando secuencialmente los modelos se reduce el bias. Por lo tanto, cada una de las estrategias reduce una parte del error total.

Forma en que se introducen variaciones en los modelos que forman el ensemble. En bagging, cada modelo es distinto del resto porque cada uno se entrena con una muestra distinta obtenida mediante bootstrapping. En boosting, los modelos se ajustan secuencialmente y la importancia (peso) de las observaciones va cambiando en cada iteración, dando lugar a diferentes ajustes.

La clave para que los métodos de ensemble consigan mejores resultados que cualquiera de sus modelos individuales es que, los modelos que los forman, sean lo más diversos posibles (sus errores no estén correlacionados). Una analogía que refleja este concepto es la siguiente: supóngase un juego como el trivial en el que los equipos tienen que acertar preguntas sobre temáticas diversas. Un equipo formado por muchos jugadores, cada uno experto en un tema distinto, tendrá más posibilidades de ganar que un equipo formado por jugadores expertos en un único tema o por un único jugador que sepa un poco de todos los temas.

A continuación, se describe con más detalle la estrategia de bagging, sobre la que se fundamenta el modelo Random Forest.

Bagging

El término bagging es el diminutivo de bootstrap aggregation, y hace referencia al empleo del muestreo repetido con reposición bootstrapping con el fin de reducir la varianza de algunos modelos de aprendizaje estadístico, entre ellos los basados en árboles.

Dadas n muestras de observaciones independientes Z1,...Zn, cada una con varianza σ2, la varianza de la media de las observaciones Z¯ es σ2/n.

En otras palabras, promediando un conjunto de observaciones se reduce la varianza.

Basándose en esta idea, una forma de reducir la varianza y aumentar la precisión de un método predictivo es obtener múltiples muestras de la población, ajustar un modelo distinto con cada una de ellas, y hacer la media (la moda en el caso de variables cualitativas) de las predicciones resultantes.

Como en la práctica no se suele tener acceso a múltiples muestras, se puede simular el proceso recurriendo a bootstrapping, generando así pseudo-muestras con los que ajustar diferentes modelos y después agregarlos. A este proceso se le conoce como bagging y es aplicable a una gran variedad de métodos de regresión.

En el caso particular de los árboles de decisión, dada su naturaleza de bajo bias y alta varianza, bagging ha demostrado tener muy buenos resultados. La forma de aplicarlo es:

Generar BB pseudo-training sets mediante bootstrapping a partir de la muestra de entrenamiento original.

Entrenar un árbol con cada una de las BB muestras del paso

Cada árbol se crea sin apenas restricciones y no se somete a pruning, por lo que tiene varianza alta pero poco bias. En la mayoría de casos, la única regla de parada es el número mínimo de observaciones que deben tener los nodos terminales. El valor óptimo de este hiperparámetro puede obtenerse comparando el out of bag error o por validación cruzada.

Para cada nueva observación, obtener la predicción de cada uno de los BB árboles. El valor final de la predicción se obtiene como la media de las BB predicciones en el caso de variables cuantitativas y como la clase predicha más frecuente (moda) para variables cualitativas.

En el proceso de bagging, el número de árboles creados no es un hiperparámetro crítico en cuanto a que, por mucho que se incremente el número, no se aumenta el riesgo de overfitting. Alcanzado un determinado número de árboles, la reducción de test error se estabiliza. A pesar de ello, cada árbol ocupa memoria, por lo que no conviene almacenar más de los necesarios.

Entrenamiento de Random Forest

El algoritmo de Random Forest es una modificación del proceso de bagging que consigue mejorar los resultados gracias a que decorrelaciona aún más los árboles generados en el proceso.

Recordando el apartado anterior, los beneficios del bagging se basan en el hecho de que, promediando un conjunto de modelos, se consigue reducir la varianza. Esto es cierto siempre y cuando los modelos agregados no estén correlacionados. Si la correlación es alta, la reducción de varianza que se puede lograr es pequeña.

Suponendo un set de datos en el que hay un predictor muy influyente, junto con otros moderadamente influyentes. En este escenario, todos o casi todos los árboles creados en el proceso de bagging estarán dominados por el mismo predictor y serán muy parecidos entre ellos. Como consecuencia de la alta correlación entre los árboles, el proceso de bagging apenas conseguirá disminuir la varianza y, por lo tanto, tampoco mejorar el modelo.

Random forest evita este problema haciendo una selección aleatoria de mm predictores antes de evaluar cada división. De esta forma, un promedio de (p−m)/p divisiones no contemplarán el predictor influyente, permitiendo que otros predictores puedan ser seleccionados. Añadiendo este paso extra se consigue decorrelacionar los árboles todavía más, con lo que su agregación consigue una mayor reducción de la varianza.

Los métodos de random forest y bagging siguen el mismo algoritmo con la única diferencia de que, en random forest, antes de cada división, se seleccionan aleatoriamente m predictores. La diferencia en el resultado dependerá del valor m escogido. Si m=p los resultados de random forest y bagging son equivalentes. Algunas recomendaciones son:

La raíz cuadrada del número total de predictores para problemas de clasificación. m≈p–√

Un tercio del número de predictores para problemas de regresión m≈p3.

Si los predictores están muy correlacionados, valores pequeños de m consiguen mejores resultados.

Sin embargo, la mejor forma para encontrar el valor óptimo de mm es evaluar el out-of-bag-error o recurrir a validación cruzada.

Al igual que ocurre con bagging, random forest no sufre problemas de overfit por aumentar el número de árboles creados en el proceso. Alcanzado un determinado número, la reducción del error de test se estabiliza.

Por otra parte como lo menciona mejor formar de evaluar los datos podría ser utilizar diversos arboles de decisión, así hacer una mejor predicción calculando el promedio de sus predicciones, este enfoque se denomina como algoritmos de ensamble o ensemble learning.

De este algoritmo se pudiera decir, que une multiples arboles de decision, así crea un bosque de predicción, las evalúa y entrega el resultado promedio. (Veloso 2019)

# 4 Desarrollo
Para trabajar con código Python, se deben cargan las librerías de Python previamente instaladas con la función py_install() de la librería reticulate de R.

La función repl_python() se utilizar para ejecutar ventana de comando o shell de Python.

Se recomienda instalar estos paquetes de Python

- py_install(packages = “pandas”)
- py_install(packages = “matplotlib”)
- py_install(packages = “numpy”)
- py_install(packages = “sklearn”) en R cloud
- py_install(“scikit-learn”) R Studio local
- py_install(packages = “statsmodels.api”)
- py_install(packages = “seaborn”)

En terminal de Python se puede actualizar con conda create -n py3.8 python=3.8 scikit-learn pandas numpy matplotlib

## 4.1 Cargar librerías


In [2]:
# Tratamiento de datos
import numpy as np
import pandas as pd
# Gráficos
import matplotlib.pyplot as plt
# Preprocesado y moYdelado
from sklearn.model_selection import train_test_split
# from sklearn.tree import DecisionTreeRegressor
# from sklearn.tree import plot_tree
# from sklearn.tree import export_graphviz
# from sklearn.tree import export_text
# from sklearn.model_selection import GridSearchCV

from sklearn.ensemble import RandomForestRegressor

from sklearn import metrics
from sklearn.metrics import mean_squared_error, r2_score

## 4.2 Cargar datos

In [3]:
datos = pd.read_csv("https://raw.githubusercontent.com/rpizarrog/Analisis-Inteligente-de-datos/main/datos/Advertising_Web.csv")
datos

Unnamed: 0    X     TV  Radio  Newspaper         Web  Sales
0             1    1  230.1   37.8       69.2  306.634752   22.1
1             2    2   44.5   39.3       45.1  302.653070   10.4
2             3    3   17.2   45.9       69.3   49.498908    9.3
3             4    4  151.5   41.3       58.5  257.816893   18.5
4             5    5  180.8   10.8       58.4  195.660076   12.9
..          ...  ...    ...    ...        ...         ...    ...
195         196  196   38.2    3.7       13.8  248.841073    7.6
196         197  197   94.2    4.9        8.1  118.041856    9.7
197         198  198  177.0    9.3        6.4  213.274671   12.8
198         199  199  283.6   42.0       66.2  237.498063   25.5
199         200  200  232.1    8.6        8.7  151.990733   13.4

[200 rows x 7 columns]

## 4.3 Explorar datos

In [4]:
print("Observaciones y variables: ", datos.shape)

Observaciones y variables:  (200, 7)


In [6]:
print("Columnas y tipo de dato")
datos.columns

Columnas y tipo de dato


Index(['Unnamed: 0', 'X', 'TV', 'Radio', 'Newspaper', 'Web', 'Sales'], dtype='object')

In [7]:
datos.dtypes

Unnamed: 0      int64
X               int64
TV            float64
Radio         float64
Newspaper     float64
Web           float64
Sales         float64
dtype: object

In [8]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  200 non-null    int64  
 1   X           200 non-null    int64  
 2   TV          200 non-null    float64
 3   Radio       200 non-null    float64
 4   Newspaper   200 non-null    float64
 5   Web         200 non-null    float64
 6   Sales       200 non-null    float64
dtypes: float64(5), int64(2)
memory usage: 11.1 KB


Se describen las variables independientes: TV, Radio Newpaper y la variable dependiente Sales.

Valor de etiqueta o variable objetivo deendiente(ventas): que significa el volumen de ventas del producto correspondiente

Las variables independientes: (TV, Radio, Periódico, WEB):

- TV: son los costos de la publicidad en TV (en miles)
- Radio: costos de publicidad invertidos en medios de difusión radio;
- Newspaper Periódico: costos publicitarios para medios impresos.
- Web: Costos de publicidad invertidos en herramientas digitales.

## 4.4 Limpiar datos
Quitar las primeras columnas y dejar TV Radio NewsPaper Web y Sales

In [9]:
datos = datos[['TV','Radio', 'Newspaper', 'Web', 'Sales']]
datos.describe()

TV       Radio   Newspaper         Web       Sales
count  200.000000  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000  159.587355   14.022500
std     85.854236   14.846809   21.778621   76.815266    5.217457
min      0.700000    0.000000    0.300000    4.308085    1.600000
25%     74.375000    9.975000   12.750000   99.048767   10.375000
50%    149.750000   22.900000   25.750000  156.862154   12.900000
75%    218.825000   36.525000   45.100000  212.311848   17.400000
max    296.400000   49.600000  114.000000  358.247042   27.000000

In [10]:
datos

TV  Radio  Newspaper         Web  Sales
0    230.1   37.8       69.2  306.634752   22.1
1     44.5   39.3       45.1  302.653070   10.4
2     17.2   45.9       69.3   49.498908    9.3
3    151.5   41.3       58.5  257.816893   18.5
4    180.8   10.8       58.4  195.660076   12.9
..     ...    ...        ...         ...    ...
195   38.2    3.7       13.8  248.841073    7.6
196   94.2    4.9        8.1  118.041856    9.7
197  177.0    9.3        6.4  213.274671   12.8
198  283.6   42.0       66.2  237.498063   25.5
199  232.1    8.6        8.7  151.990733   13.4

[200 rows x 5 columns]

## 4.5 Datos de entrenamiento y datos de validación
Se utiliza semilla 1550 (random_state=1550)

La función train_test_split() parte los datos originales el 70% y 30% para datos de entrenamiento y validación y con el argumento datos.drop(columns = “Sales”), datos[‘Sales’] solo incluye las variables independientes; la semilla de aleatoriedad es 2022.

In [11]:
X_entrena, X_valida, Y_entrena, Y_valida = train_test_split(datos.drop(columns = "Sales"), datos['Sales'],train_size=.70,  random_state=1550)

### 4.5.1 Datos de entrenamiento

In [12]:
print("Estructura de datos de entrenamiento... ", X_entrena.shape)

Estructura de datos de entrenamiento...  (140, 4)


In [13]:
print(X_entrena)

        TV  Radio  Newspaper         Web
53   182.6   46.2       58.7  176.050052
145  140.3    1.9        9.0  231.883385
54   262.7   28.8       15.9  324.615179
90   134.3    4.9        9.3  258.355488
52   216.4   41.7       39.6  161.802512
..     ...    ...        ...         ...
15   195.4   47.7       52.9  148.095134
181  218.5    5.4       27.4  162.387486
137  273.7   28.9       59.7  288.260611
25   262.9    3.5       19.5  160.562859
74   213.4   24.6       13.1  156.284261

[140 rows x 4 columns]


In [14]:
print(X_entrena[['TV']], X_entrena[['Radio']], X_entrena[['Newspaper']], X_entrena[['Web']])

        TV
53   182.6
145  140.3
54   262.7
90   134.3
52   216.4
..     ...
15   195.4
181  218.5
137  273.7
25   262.9
74   213.4

[140 rows x 1 columns]      Radio
53    46.2
145    1.9
54    28.8
90     4.9
52    41.7
..     ...
15    47.7
181    5.4
137   28.9
25     3.5
74    24.6

[140 rows x 1 columns]      Newspaper
53        58.7
145        9.0
54        15.9
90         9.3
52        39.6
..         ...
15        52.9
181       27.4
137       59.7
25        19.5
74        13.1

[140 rows x 1 columns]             Web
53   176.050052
145  231.883385
54   324.615179
90   258.355488
52   161.802512
..          ...
15   148.095134
181  162.387486
137  288.260611
25   160.562859
74   156.284261

[140 rows x 1 columns]


## 4.6 Modelo de Random Forest
Se crea el modelo con la función RandomForestRegressor().fit() con los datos de entrenamiento que contiene la variables independientes en relación a la variable dependiente ‘Sales’.

El argumento n_estimators define el número de árboles en el bosque aleatorio. Puede usar cualquier valor numérico para el parámetro n_estimator.

El valor argumento de random_state es la semilla para la aleatoriedad del modelo

In [15]:
model_rf = RandomForestRegressor(n_estimators = 50, random_state = 1550)

model_rf.fit(X_entrena, Y_entrena)

RandomForestRegressor(n_estimators=50, random_state=1550)

## 4.7 Construir predicciones
Se generan predicciones con los datos de validación

In [21]:
predicciones = model_rf.predict(X_valida)
predicciones

array([ 4.792,  6.188, 23.48 ,  9.436, 18.672, 20.096, 23.98 , 11.994,
       24.922, 19.842, 19.422, 10.752, 10.702, 22.316, 18.132,  6.92 ,
        6.978, 19.234, 22.488, 24.   ,  7.38 , 15.614, 13.41 ,  9.922,
       15.928, 12.154,  9.676, 13.224, 18.6  , 11.966, 10.704, 21.476,
       17.62 , 13.906, 11.376, 12.09 , 15.618, 22.462, 18.912, 11.87 ,
       15.384, 13.362, 13.202,  6.566, 11.82 , 13.222, 15.81 ,  9.448,
       21.53 ,  5.038, 10.29 , 14.232, 11.008, 17.808,  9.44 , 11.956,
        9.63 ,  7.576, 12.832, 10.792])

## 4.8 Construir matriz de comparación
Construir una conjunto de datos con los valores reales de los datos de validación y las predicciones generadas del modelo.

In [17]:
comparaciones = pd.DataFrame(Y_valida)

# comparaciones = comparaciones.assign(Sales_Real = Y_valida)

comparaciones = comparaciones.assign(Predicho = predicciones.flatten().tolist())

print(comparaciones)

     Sales  Predicho
155    3.2     4.792
5      7.2     6.188
17    24.4    23.480
107    8.7     9.436
28    18.9    18.672
176   20.2    20.096
61    24.2    23.980
81    12.3    11.994
183   26.2    24.922
141   19.2    19.842
133   19.6    19.422
18    11.3    10.752
157   10.1    10.702
55    23.7    22.316
41    17.1    18.132
192    5.9     6.920
189    6.7     6.978
142   20.1    19.234
58    23.8    22.488
198   25.5    24.000
91     7.3     7.380
154   15.6    15.614
37    14.7    13.410
65     9.3     9.922
45    14.9    15.928
146   13.2    12.154
32     9.6     9.676
114   14.6    13.224
153   19.0    18.600
151   11.6    11.966
6     11.8    10.704
139   20.7    21.476
105   19.2    17.620
179   12.6    13.906
144   11.4    11.376
100   11.7    12.090
19    14.6    15.618
111   21.8    22.462
109   19.8    18.912
174   11.5    11.870
171   14.5    15.384
7     13.2    13.362
31    11.9    13.202
22     5.6     6.566
131   12.7    11.820
116   12.2    13.222
123   15.2   

## 4.9 Evaluación del modelo con RMSE
Este valor normalmente se compara contra otro modelo y el que esté mas cerca de cero es mejor.

La raiz del Error Cuadrático Medio (rmse) es una métrica que dice qué tan lejos están los valores predichos de los valores observados o reales en un análisis de regresión, en promedio. 

RMSE es una forma útil de ver qué tan bien un modelo de regresión puede ajustarse a un conjunto de datos.

Cuanto mayor sea el rmse, mayor será la diferencia entre los valores predichos y reales, lo que significa que peor se ajusta un modelo de regresión a los datos. Por el contrario, cuanto más pequeño sea el rmse, mejor podrá un modelo ajustar los datos.

Se compara este valor de rmse con respecto al modelo de regresión múltiple



In [18]:
rmse = mean_squared_error(
        y_true  = Y_valida,
        y_pred  = predicciones,
        squared = False
       )
print(f"El error (rmse) de test es: {rmse}")

El error (rmse) de test es: 0.839761791621091


In [19]:
print('Mean Squared Error: MSE', metrics.mean_squared_error(Y_valida, predicciones))


Mean Squared Error: MSE 0.7051998666666647


In [20]:
print('Root Mean Squared Error RMSE:', np.sqrt(metrics.mean_squared_error(Y_valida, predicciones)))


Root Mean Squared Error RMSE: 0.839761791621091


Para estos datos para estos valores de entrenamiento y validación, con un valor de 50 árboles en el modelo, el valor de RMSE es aproximadamente 0.7660, menor que el valor de árbol de regresión en Python del Caso 4.

El RMSE obtenido se puede comparar contra los estadísticos RMSE de los modelos regresión múltiple y árbol de regresión tanto de Python y de R , se puede interpretar cuál modelo es más eficiente en términos de este estadístico con estos datos.

# Interpretación
Comparando los valores RMSE obtenidos tanto en RStudio como en Python, concluyo que el modelo de Python es mucho más eficaz que aquel de RStudio, esto debido a que la medida RMSE es mucho menor, siendo esta un ~0.839761791621091, la cual es una buena medida para el RMSE.